Status color of glyphs

ollimeier opened the pull request:
Fixes #926

Here a screenshot:
<img width="1579" alt="Screenshot 2024-05-07 at 11 44 28" src="https://github.com/googlefonts/fontra/assets/21055547/1ccfd54a-ca2a-4b7f-926b-5b13ff48d26b">

justvanrossum reviewed (commented):
Thanks! Looks very good.

Possible refinements:
- reduce the opacity of the color while in edit mode, maybe 50%?
- while I don't think the thickness of the color bar should be fully fixed in screen pixels, I feel it is getting too thin at very small sizes, and too dominant when zooming in. Suggestion: define a minimum and maximum thickness in screen pixels, and clamp the 50 value between them. Something like this:
```js
  const thickness = clamp(50, parameters.minThickness, parameters.maxThickness);
```

justvanrossum commented on src/fontra/views/editor/visualization-layer-definitions.js:
```diff
@@ -1020,6 +1021,45 @@ registerVisualizationLayerDefinition({
+    context.beginPath();
+    context.rect(
+      0,
+      (-150 * model.fontController.unitsPerEm) / 1000,
```
```suggestion
      -0.150 * model.fontController.unitsPerEm,
```

justvanrossum commented on src/fontra/views/editor/visualization-layer-definitions.js:
```diff
@@ -1020,6 +1021,45 @@ registerVisualizationLayerDefinition({
+      0,
+      (-150 * model.fontController.unitsPerEm) / 1000,
+      positionedGlyph.glyph.xAdvance,
+      50
```
```suggestion
      0.05 * model.fontController.unitsPerEm
```

justvanrossum reviewed (commented):
- Ane minor suggestion
- The value of "50" should also be related to the units per em (but see also previous review)

ollimeier commented:
Thanks a lot for your feedback, Just. 
I added a opacity for the status color. But to make it work I had to create a new function `rgbaToFillStyle`, because `rgbaToCSS` does not work with the alpha value for canvas 2d, please see the following documentation:
[https://doc.qt.io/qt-5/qml-qtquick-context2d.html#fillStyle-prop](https://doc.qt.io/qt-5/qml-qtquick-context2d.html#fillStyle-prop)
Or do you have a different idea how to solve this?

Also, to make it work with two different colors (with and without transparency) for editing and non-editing mode had to split it into two separate `registerVisualizationLayerDefinition` or is there a way to make it possible with one, which I am not aware of?

Beside that, I tried to implement `const thickness = clamp(50, parameters.minThickness, parameters.maxThickness);`, but when I added minThickness and maxThickness to the parameters, the values were scaled (related to the zoom), resulting in weird values/sizes for the status color box in the background.


I hope this all make somehow sense to you.
Here a short demo video of the modifications:

https://github.com/googlefonts/fontra/assets/21055547/1d635f46-2844-4a61-9e49-2a30d2a4235a

justvanrossum reviewed (commented):
> I added a opacity for the status color. But to make it work I had to create a new function rgbaToFillStyle, because rgbaToCSS does not work with the alpha value for canvas 2d, please see the following documentation:
https://doc.qt.io/qt-5/qml-qtquick-context2d.html#fillStyle-prop

It is not that rgbaToCSS() does not work with the alpha value for canvas 2d, it does not work with the alpha value, period. You have discovered a bug :) So instead of adding a new function, you should fix the bug in the existing function.

> Also, to make it work with two different colors (with and without transparency) for editing and non-editing mode had to split it into two separate registerVisualizationLayerDefinition or is there a way to make it possible with one, which I am not aware of?

Good point. While it is possible to compare positionedGlyph to model.getSelectedPositionedGlyph(), this is not ideal, and I have created https://github.com/googlefonts/fontra/pull/1345 to make this easier. Rebase to main, and you can use the positionedGlyph.isEditing flag.

> Beside that, I tried to implement const thickness = clamp(50, parameters.minThickness, parameters.maxThickness);, but when I added minThickness and maxThickness to the parameters, the values were scaled (related to the zoom), resulting in weird values/sizes for the status color box in the background.

Having them scaled was exactly my intention. I want min and max values that are in pixels ("let the status bar not be thinner than X pixels and not be thicker than Y pixels"). So, calculate a base thickness by doing 0.05 * unitsPerEm (*not* scaled to pixels, this is in font units), and clamp that value between the *scaled* min and max values. Suggested minThickness: 3; suggested maxThickness: 15. We can fine tune later. This is exactly what I described earlier, though, but I hope it is clearer now.

ollimeier commented on src/fontra/views/editor/visualization-layer-definitions.js:
```diff
@@ -1020,6 +1022,60 @@ registerVisualizationLayerDefinition({
+
+  const thickness = clamp(parameters.thickness, 20, 150);
+  let color = statusFieldDefinitions[status].color;
+  color[3] = parameters.opacity;
```
Thank you very much. I very much hope it's better now.

ollimeier commented on src/fontra/client/core/utils.js:
```diff
@@ -352,6 +352,22 @@ export function rgbaToCSS(rgba) {
 
+// according to the documentation, the alpha value must be a value
+// between 0.0 and 1.0, why it's not possible to use rgbaToCSS
+// more: https://doc.qt.io/qt-5/qml-qtquick-context2d.html#fillStyle-prop
```
I reverted all my changes and reworked it based on your feedback. For example, this new function has been removed.

justvanrossum commented on src/fontra/client/core/utils.js:
```diff
@@ -345,9 +345,10 @@ export function arrayExtend(thisArray, itemsArray) {
-  if (channels[3] == 255) {
-    channels.pop();
+  const channels = rgba.slice(0, 3).map((channel) => Math.round(channel * 255));
+  const alpha = rgba.length === 4 ? rgba[3] : undefined;
```
In JS, you can just ask for out-of-bounds indices and get undefined:
```suggestion
  const alpha = rgba[3];
```

ollimeier replied:
Thank you. I changed it. Please see #1346

justvanrossum commented on src/fontra/views/editor/visualization-layer-definitions.js:
```diff
@@ -1020,6 +1022,55 @@ registerVisualizationLayerDefinition({
+      parameters.maxThickness
+    );
+    context.fillStyle = rgbaToCSS(color);
+    context.beginPath();
```
There is `context.fillRect()` as a shortcut for these three lines.

ollimeier replied:
Thank you. I changed it. Please see #1346

ollimeier commented:
I am closing this PR, because there are some weird conflicts. I will create a new one in a sec.

justvanrossum commented on src/fontra/views/editor/visualization-layer-definitions.js:
```diff
@@ -1020,6 +1022,55 @@ registerVisualizationLayerDefinition({
+    const color = [...statusFieldDefinitions[status].color];
+    if (positionedGlyph.isEditing) {
+      // in editing mode make opacity 50%
+      color[3] = color[3] * 0.5;
```
```suggestion
      color[3] = color[3] * 0.4;
```

ollimeier replied:
Thank you. I changed it. Please see #1346

justvanrossum reviewed (commented):
Some minor comments. Almost done!
